src/NServiceBus.Core/Recoverability/FailureInfoStorage.cs

```diff
@@ -0,0 +1,122 @@
+        }
+
+        public void MarkForMovingToErrorQueue(string messageId, ExceptionDispatchInfo exceptionDispatchInfo)
+        {
+            RecordFailureInfoForMessage(messageId, exceptionDispatchInfo, true, false);
+        }
+
+        public void MarkForDeferralForSecondLevelRetry(string messageId, ExceptionDispatchInfo exceptionDispatchInfo)
+        {
+            RecordFailureInfoForMessage(messageId, exceptionDispatchInfo, false, true);
+        }
+
+        void RecordFailureInfoForMessage(string messageId, ExceptionDispatchInfo exceptionDispatchInfo, bool shouldMoveToErrorQueue, bool shouldDeferForSecondLevelRetry)
+        {
+            lock (lockObject)
+            {
+                FailureInfoNode node;
+                if (failureInfoPerMessage.TryGetValue(messageId, out node))
+                {
+                    // We have seen this message before, just update the counter and store exception.
+                    node.FailureInfo = new ProcessingFailureInfo(
+                        NumberOfFailedFlrAttempts(shouldMoveToErrorQueue, shouldDeferForSecondLevelRetry, node),
+                        exceptionDispatchInfo,
+                        shouldMoveToErrorQueue,
+                        shouldDeferForSecondLevelRetry);
+
+                    // Maintain invariant: leastRecentlyUsedMessages.First contains the LRU item.
+                    leastRecentlyUsedMessages.Remove(node.LeastRecentlyUsedEntry);
+                    leastRecentlyUsedMessages.AddLast(node.LeastRecentlyUsedEntry);
+                }
+                else
+                {
+                    if (failureInfoPerMessage.Count == maxElements)
+                    {
+                        // We have reached the maximum allowed capacity. Remove the LRU item.
+                        var leastRecentlyUsedEntry = leastRecentlyUsedMessages.First;
+                        failureInfoPerMessage.Remove(leastRecentlyUsedEntry.Value);
+                        leastRecentlyUsedMessages.RemoveFirst();
+                    }
+
+                    var newNode = new FailureInfoNode(
+                        messageId,
+                        new ProcessingFailureInfo(1, exceptionDispatchInfo, shouldMoveToErrorQueue, shouldDeferForSecondLevelRetry));
+
+                    failureInfoPerMessage[messageId] = newNode;
+
+                    // Maintain invariant: leastRecentlyUsedMessages.First contains the LRU item.
+                    leastRecentlyUsedMessages.AddLast(newNode.LeastRecentlyUsedEntry);
+                }
+            }
+        }
+
+        static int NumberOfFailedFlrAttempts(bool shouldMoveToErrorQueue, bool shouldDeferForSecondLevelRetry, FailureInfoNode node)
+        {
+            var isFlrAttemptFailure = !shouldMoveToErrorQueue && !shouldDeferForSecondLevelRetry;
+            if (isFlrAttemptFailure)
+            {
+                return node.FailureInfo.NumberOfFailedFlrAttempts + 1;
```

Comment:
FLR counter should not be increased by one by the storage. The value should be passed from the behavior which encapsulates the logic for changing that value.